### PR TITLE
Issue #2866: added prefilters and exclusion prefilter

### PR DESCRIPTION
**Pull request by rnveach**
Issue #2866

Added pre-filters to be able to exclude files from being checked.
ExclusionPreFilter accepts regular expression to allow customizing which files are skipped.

Missing any XDoc documentation as there is some question if `PreFilter` is the word we want to use to describe this new set.

Currently prefilters is missing from `Checker.getExternalResourceLocations` as the only pre-filter doesn't use this and requirements in issue didn't want this.
If we still want pre-filters to have external resources, then I can just add a dummy class to fill in the code coverage.

### CODE REVIEW

**Comment by rnveach** [comment]
@romani How do we please TC and PMD?
PMD gives the error: `[INFO] PMD Failure: com\puppycrawl\tools\checkstyle\Checker.java:63 Rule:TooManyFields Priority:3 Too many fields.`
But TC doesn't like the PMD suppression and gives `Inspection suppression annotation @SuppressWarnings("PMD.TooManyFields")`

**Comment by romani** [comment]
@rnveach ,

> How do we please TC and PMD?

all PMD excludes are hosted at https://github.com/checkstyle/checkstyle/blob/master/config/pmd.xml (Example: https://github.com/checkstyle/checkstyle/blob/master/config/pmd.xml#L112 ).
The same for all other analysis tools.
I still do not like an approach to suppress warnings of some tools in sources. I still think that it is a pollution of sources. IntelijIdea inspections have an some sort of excuse only because suppression could be placed to javadoc that is not visible during reading of code.

> And I don't see a way right now to pull out the fields.

Lets keep it in excludes for now. I do not know how to split Checker for now.

**Comment by rnveach** [comment]
@romani Thanks that fixed the issue. I missed that file.
Documentation is still missing. If your ok with implementation/interface and naming, I will begin writing up the documentation.

**Comment by romani** [comment]
@rnveach , sorry for delay in reply ....

1)
> ExclusionPreFilter , PreFilterSet, PreFilter

without documentation it looks like a class that is running before Filters.

Lets make them more distinctive - `BeforeExecutionExclusionFilter`, `BeforeExecutionFilterSet`, `BeforeExecutionFilter`
or
`BeforeExecutionExclusionFileFilter`, `BeforeExecutionFileFilterSet`, `BeforeExecutionFileFilter`
(filter term is already use AuditEvent, so FileFilter will make it clear that filter is different)

Names a bit long but self descriptive, might not require documentation reading.

do you have better names ?

2)
> public class PreFilterSet
   public boolean equals(Object obj) {
   public int hashCode() {

>public class ExclusionPreFilter
public boolean equals(Object obj) {
public int hashCode() {

Do we need this ? if yes, I would rather make class as immutable and even final, to not let anybody be able to misuse/lost it in hashmap (object mutation after it is appeared in hashmap collection).

**Comment by romani** [comment]
@rnveach , issues need to be resolved, and rebased to resolve conflicts

**Comment by rnveach** [comment]
Sorry for the delay.

1) I don't, I will go with `BeforeExecutionExclusionFileFilter`... I named package `filefilters` so it is not as long as class name.
2) This was a strict copy/paste from `FilterSet`/`SuppressionFilter` since the 2 classes have similar purposes.
I removed them and made the classes final. `ExclusionPreFilter` can not be immutable since it is a bean and we populate the fields after the instance is created.
3)
Here is a report using `checks-only-javadoc-error.xml` on JDK9's master using the additional lines:
````
<module name="BeforeExecutionExclusionFileFilter">
<property name="fileName" value="module\-info\.java|[\\/]src[\\/]main[\\/]java[\\/]test[\\/]"/>
</module>
````
http://rveach.no-ip.org/checkstyle/regression/reports/85/ (73k warnings).

I had to exclude the test directory because I was getting too many parse errors from other files. I will create an issue for them all when I get the chance.
I did multiple runs with different exclusions, so I believe the code is working as desired.
I did not create the documentation yet, that is the next thing I will work on.

**Comment by rnveach** [comment]
@romani Documentation was added.

**Comment by romani** [comment]
1)
>  I will create an issue for them all when I get the chance.

please reference files that we cannot still parse in jdk9.

2)
please create PR to https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/projects-for-circle.properties#L8 to minimize exclusions and use HEAD code.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/BeforeExecutionFileFilter.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * An interface for before execution file filtering events.
+ * @author Richard Veach
+ */
+public interface BeforeExecutionFileFilter {
+    /**
+     * Determines whether or not a before execution file filtered event is accepted.
+     * @param event the BeforeExecutionFileFilterEvent to filter.
+     * @return true if the event is accepted.
+     */
+    boolean accept(BeforeExecutionFileFilterEvent event);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/BeforeExecutionFileFilter.java]: Do we really need BeforeExecutionFileFilterEvent ?

Why not simple `boolean accept(String uri);` ? What else xxxxxxxxxxxFileFilter could care about ?
https://prateekvjoshi.com/2014/02/22/url-vs-uri-vs-urn/ . We just need string representation of location.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/BeforeExecutionFileFilterSet.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    @Override
+    public boolean accept(BeforeExecutionFileFilterEvent event) {
+        for (BeforeExecutionFileFilter filter : beforeExecutionFileFilters) {
+            if (!filter.accept(event)) {
+                return false;
+            }
+        }
+        return true;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/BeforeExecutionFileFilterSet.java]: single return point please

File: src/test/java/com/puppycrawl/tools/checkstyle/DebugBeforeExecutionFileFilter.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+import com.puppycrawl.tools.checkstyle.api.BeforeExecutionFileFilter;
+import com.puppycrawl.tools.checkstyle.api.BeforeExecutionFileFilterEvent;
+
+class DebugBeforeExecutionFileFilter implements BeforeExecutionFileFilter {
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/DebugBeforeExecutionFileFilter.java]: I think it is reasonable to name it "TestBeforeExecutionFileFilter"

**Comment by rnveach** [comment]
1) Files are referenced here: https://github.com/checkstyle/checkstyle/issues/3033#issuecomment-252448394
2) Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/BeforeExecutionFileFilter.java (hunk 1 above)
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/BeforeExecutionFileFilter.java]: > Why not simple boolean accept(String uri); ?

Again, another copy/paste from FilterSet/SuppressionFilter.

> What else xxxxxxxxxxxFileFilter could care about ?

Don't forget the event also has the [source](https://github.com/checkstyle/checkstyle/pull/3386/files/60e22ceb5e117000d8c0c8cdce3b9871ca83cf57#diff-72cad0b13e0214053c4ceef4a803e9dbR42), though we don't use it even in `AuditEvent`s. I assume there is a reason we went with it originally.

Done.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/BeforeExecutionFileFilterSet.java (hunk 2 above)
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/BeforeExecutionFileFilterSet.java]: Again, another copy/paste from FilterSet/SuppressionFilter.
Done.

File: src/test/java/com/puppycrawl/tools/checkstyle/DebugBeforeExecutionFileFilter.java (hunk 3 above)
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/DebugBeforeExecutionFileFilter.java]: Again, another copy/paste from FilterSet/SuppressionFilter.
Done.

**Pull request merged**